src/lib/expression/evaluation/expression_evaluator.cpp

```diff
@@ -0,0 +1,1054 @@
+    }
+
+    end = start + length;
+    start = std::max(0, start);
+    end = std::min(end, signed_string_size);
+    length = end - start;
+
+    // Invalid/out of range arguments, like SUBSTR("HELLO", 4000, -2), lead to an empty string
+    if (!string.empty() && start >= 0 && start < signed_string_size && length > 0) {
+      length = std::min<int32_t>(signed_string_size - start, length);
+      result_values[chunk_offset] = string.substr(static_cast<size_t>(start), static_cast<size_t>(length));
+    }
+  }
+
+  return std::make_shared<ExpressionResult<std::string>>(result_values, result_nulls);
+}
+
+std::shared_ptr<ExpressionResult<std::string>> ExpressionEvaluator::_evaluate_concatenate(
+    const std::vector<std::shared_ptr<AbstractExpression>>& arguments) {
+  std::vector<std::shared_ptr<ExpressionResult<std::string>>> argument_results;
+  argument_results.reserve(arguments.size());
+
+  auto result_is_nullable = false;
+
+  // I - Compute the arguments
+  for (const auto& argument : arguments) {
+    // CONCAT with a NULL literal argument -> result is NULL
+    if (argument->data_type() == DataType::Null) {
+      auto null_value_result = ExpressionResult<std::string>{{{}}, {true}};
+      return std::make_shared<ExpressionResult<std::string>>(null_value_result);
+    }
+
+    const auto argument_result = evaluate_expression_to_result<std::string>(*argument);
+    argument_results.emplace_back(argument_result);
+
+    result_is_nullable |= argument_result->is_nullable();
+  }
+
+  // II - Compute the number of output rows
+  auto result_size = argument_results.empty() ? size_t{0} : argument_results.front()->size();
+  for (auto argument_idx = size_t{1}; argument_idx < argument_results.size(); ++argument_idx) {
+    result_size = _result_size(result_size, argument_results[argument_idx]->size());
+  }
+
+  // II - Concatenate the values
+  std::vector<std::string> result_values(result_size);
+  for (const auto& argument_result : argument_results) {
+    argument_result->as_view([&](const auto& argument_view) {
+      for (auto chunk_offset = ChunkOffset{0}; chunk_offset < result_size; ++chunk_offset) {
+        // The actual CONCAT
+        result_values[chunk_offset] += argument_view.value(chunk_offset);
+      }
+    });
+  }
+
+  // III - Optionally concatenate the nulls and return
+  if (result_is_nullable) {
+    std::vector<bool> result_nulls(result_size, false);
+    for (const auto& argument_result : argument_results) {
+      argument_result->as_view([&](const auto& argument_view) {
+        for (auto chunk_offset = ChunkOffset{0}; chunk_offset < result_size; ++chunk_offset) {
+          result_nulls[chunk_offset] = result_nulls[chunk_offset] || argument_view.null(chunk_offset);
+        }
+      });
+    }
+
+    return std::make_shared<ExpressionResult<std::string>>(std::move(result_values), std::move(result_nulls));
+  } else {
+    return std::make_shared<ExpressionResult<std::string>>(std::move(result_values));
+  }
+}
+
```

Comment:
Attaching this somewhere at random:
```
(debug)> load src/test/tables/int_float_with_null.tbl tbl2
Loading src/test/tables/int_float_with_null.tbl into table "tbl2" ...
(debug)> INSERT INTO tbl2 VALUES (1,2);
src/lib/storage/value_column.cpp:130 This ValueColumn does not support null values.
(debug)!> INSERT INTO tbl2 VALUES (1,2.9);
libc++abi.dylib: terminating with uncaught exception of type std::logic_error: src/lib/concurrency/transaction_context.cpp:32 A registered operator failed but transaction has not been rolled back.
```
In the current master, we would also fail, but do so at the right time and loudly:
```
(debug)> INSERT INTO tbl2 VALUES (1,2);
src/lib/sql/sql_translator.cpp:198 Insert: Column type mismatch
Segmentation fault: 11
```

Reply:
will address

Reply:
The fault here lies with the Insert operator, imo.
`This ValueColumn does not support null values` comes from it trying to insert a NULL (derived from the 2-which-is-not-a-float) into the target column, even if it is not nullable. Let's discuss Insert here #972?